### PR TITLE
Gh 15/safe selector

### PR DESCRIPTION
**Pull request by MouazAlzahabi**
feat: Safe Selector (#15 )

### CODE REVIEW

File: Multisig/UI/Safe Management/SafeInfoView.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                }) {
+                    Image("icon-external-link")
+                }.foregroundColor(.gnoHold).frame(width: 24, height: 24)
+            }
+            .padding(.leading, 60)
+            .padding(.trailing, 24)
+            Text(safe?.ensName ?? "")
+            .font(Font.gnoBody.weight(.medium))
+            .multilineTextAlignment(.center)
+            QRView(value: safe?.address ?? "").frame(width: 124, height: 124)
```
Comment (DmitryBespalov) [inline Multisig/UI/Safe Management/SafeInfoView.swift:42]: please separate each element with blank line, it is hard to read like it is now.

Also, the modifiers need to be indented to the right one tab

File: Multisig/UI/App/ContentView.swift [hunk 2]
```diff
@@ -10,18 +10,23 @@ import SwiftUI
...
+    @State private var showSafeSelectorSheet: Bool = false
+    @State private var activeSheet: SafeSelectorActiveOption = .none
+    @Environment(\.managedObjectContext) var context
+    
+    @FetchRequest(fetchRequest: AppSettings.settings()) var appSettings: FetchedResults<AppSettings>
     var body: some View {
-        VStack(spacing: 0) {
-
-            selector.zIndex(1)
+        return VStack(spacing: 0) {
```
Comment (sche) [inline Multisig/UI/App/ContentView.swift:19]: "return" not necessary

File: Multisig/UI/App/ContentView.swift [hunk 3]
```diff
@@ -10,18 +10,23 @@ import SwiftUI
...
+            SafeSelector(showSheet: $showSafeSelectorSheet, activeSheet: $activeSheet).zIndex(1).environment(\.managedObjectContext, context)
```
Comment (sche) [inline Multisig/UI/App/ContentView.swift:20]: Please shift methods to separate lines for readability.

File: Multisig/UI/App/ContentView.swift [hunk 4]
```diff
@@ -10,18 +10,23 @@ import SwiftUI
...
+            .sheet(isPresented: self.$showSafeSelectorSheet) {
+                //TODO: handle show safe info screen
```
Comment (sche) [inline Multisig/UI/App/ContentView.swift:22]: Do we need this todo?

Comment (MouazAlzahabi) [reply Multisig/UI/App/ContentView.swift:22]: I will remove

File: Multisig/UI/Safe Management/SafeInfoView.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @FetchRequest(fetchRequest: AppSettings.settings()) var appSettings: FetchedResults<AppSettings>
+    @Environment(\.managedObjectContext) var context: NSManagedObjectContext
+    
+    var body: some View {
+        var safe: Safe?
+        if let address = appSettings[0].selectedSafe {
+            safe = try? context.fetch(Safe.by(address: address)).first
+        }
+        
+        return VStack (alignment: .center, spacing: 18){
```
Comment (sche) [inline Multisig/UI/Safe Management/SafeInfoView.swift:22]: please remove space before "(" and add after ")"

File: Multisig/UI/Safe Management/SwitchSafeView.swift [hunk 6]
```diff
@@ -62,6 +61,7 @@ struct SwitchSafeView: View {
             .background(Rectangle().foregroundColor(.white))
             .onTapGesture {
                 self.appSettings.selectedSafe = self.safe.address
+                self.presentationMode.wrappedValue.dismiss()
```
Comment (sche) [inline Multisig/UI/Safe Management/SwitchSafeView.swift:64]: Do we need it?

File: Multisig/UI/Safe Management/SafeInfoView.swift [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+import SwiftUI
+import CoreData
+
+struct SafeInfoView: View {
+    @FetchRequest(fetchRequest: AppSettings.settings()) var appSettings: FetchedResults<AppSettings>
+    @Environment(\.managedObjectContext) var context: NSManagedObjectContext
+    
+    var body: some View {
+        var safe: Safe?
```
Comment (sche) [inline Multisig/UI/Safe Management/SafeInfoView.swift:17]: Can we provide here Safe object in the constructor, as it is already known when we call this view (I guess so, not sure)?

File: Multisig/UI/Safe Management/SwitchSafeView.swift (hunk 6 above)
Comment (MouazAlzahabi) [reply Multisig/UI/Safe Management/SwitchSafeView.swift:64]: Yes in order to dismiss the screen after selecting new safe

File: Multisig/UI/UI Library/SafeSelector.swift [hunk 8]
```diff
@@ -7,20 +7,60 @@
...
 import SwiftUI
+import CoreData
 
 struct SafeSelector: View {
     var height: CGFloat = 116
-
+    
+    @FetchRequest(fetchRequest: AppSettings.settings()) var appSettings: FetchedResults<AppSettings>
+    @Environment(\.managedObjectContext) var context: NSManagedObjectContext
+    @Binding var showSheet: Bool
```
Comment (sche) [inline Multisig/UI/UI Library/SafeSelector.swift:17]: Seems like it is enough to have it as `@State`

File: Multisig/UI/UI Library/SafeSelector.swift [hunk 9]
```diff
@@ -7,20 +7,60 @@
...
+    @Binding var activeSheet: SafeSelectorActiveOption
```
Comment (sche) [inline Multisig/UI/UI Library/SafeSelector.swift:18]: here as well

File: Multisig/UI/UI Library/SafeSelector.swift [hunk 10]
```diff
@@ -7,20 +7,60 @@
...
+                Button(action: {
+                    self.showSheet.toggle()
+                    self.activeSheet = .info
+                }) {
+                    HStack {
+                        Identicon(safe?.address ?? "").frame(width: 36, height: 36)
+                            .padding()
+                        VStack (alignment: .leading){
+                            Text(safe?.name ?? "")
+                            .font(Font.gnoBody.weight(.medium))
```
Comment (sche) [inline Multisig/UI/UI Library/SafeSelector.swift:46]: please tabulate for readability

File: Multisig/UI/UI Library/SafeSelector.swift [hunk 11]
```diff
@@ -29,8 +69,17 @@ struct SafeSelector: View {
...
 }
 
+enum SafeSelectorActiveOption {
+    case info, switchSafe, none
+}
+
 struct SafeSelector_Previews: PreviewProvider {
     static var previews: some View {
-        SafeSelector()
+//        let context = TestCoreDataStack().persistentContainer.viewContext
```
Comment (sche) [inline Multisig/UI/UI Library/SafeSelector.swift:78]: please remove commented lines

File: Multisig/UI/UI Library/ShortAddressText.swift [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    init(_ text: String, prefixCount: Int, suffixCount: Int) {
+        self.prefixCount = prefixCount
+        self.suffixCount = suffixCount
+        self.text = text
+        assert(text.count > prefixCount + suffixCount)
+    }
+
+    var body: some View {
+        Group {
+            Text(prefix).foregroundColor(bodyColor) +
```
Comment (sche) [inline Multisig/UI/UI Library/ShortAddressText.swift:35]: Will it work if we put `foregroundColor` as Group property?

File: Multisig/Logic/Models/Safe.swift [hunk 13]
```diff
@@ -55,4 +53,14 @@ extension Safe: Identifiable {
...
     }
 
+    static func selected(_ settings: AppSettings) -> Safe? {
+        if let address = settings.selectedSafe {
+            let request = Safe.by(address: address)
+            return try? CoreDataStack.shared.viewContext.fetch(request).first
+        } else {
+            return nil
+        }
+
```
Comment (sche) [inline Multisig/Logic/Models/Safe.swift:63]: extra space

File: Multisig/UI/Safe Management/SwitchSafeView.swift (hunk 6 above)
Comment (sche) [reply Multisig/UI/Safe Management/SwitchSafeView.swift:64]: will it work if we just take it from ```@Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>``` ?

**Comment by MouazAlzahabi** [comment]
feat: safe selector #15
closes #24

**Pull request merged**